### PR TITLE
ROX-22371, ROX-22372: compliance update scan config delete cluster

### PR DESCRIPTION
**Pull request by dashrews78**
## Description

When a scan configuration was updated such that it removed a cluster, we were not deleting the configuration on the deleted sensor nor were we removing the cluster from the scan configuration status.  This PR does both and refactors things slightly to separate the creation of a scan configuration from the update.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
~- [ ] Evaluated and added CHANGELOG entry if required~
~- [ ] Determined and documented upgrade steps~
~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs]~(https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

Added some unit test cases.  
Additionally did some testing running on 2 clusters to ensure that the delete was pushed to sensor.

The initial configuration with both clusters
Screenshot 2024-02-13 at 3 44 13 PM

After removing the second cluster from the scan configuration
Screenshot 2024-02-13 at 3 07 52 PM

The before and after of showing the suites running on cluster 2.
Screenshot 2024-02-13 at 3 08 08 PM

The suites that ran on cluster 1.
Screenshot 2024-02-13 at 3 08 20 PM

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: central/complianceoperator/v2/compliancemanager/manager_impl.go [hunk 1]
```diff
@@ -198,6 +232,38 @@ func (m *managerImpl) ProcessScanRequest(ctx context.Context, scanRequest *stora
 	return scanRequest, nil
 }
 
+func (m *managerImpl) processClusterDelete(ctx context.Context, scanRequest *storage.ComplianceOperatorScanConfigurationV2, clusters []string) {
+	for _, clusterID := range clusters {
+		log.Infof("SHREWS -- trying to delete cluster %q", clusterID)
+		// id for the request message to sensor
+		sensorRequestID := uuid.NewV4().String()
+
+		sensorMessage := &central.MsgToSensor{
```
Comment (SimonBaeumer) [inline central/complianceoperator/v2/compliancemanager/manager_impl.go:241]: What happens if a delete and a scan are executed at the same time?
The scan requests are tracked in `runningRequests`, but the `processClusterDelete` is not.

**Review by SimonBaeumer — APPROVED** [review]

File: central/complianceoperator/v2/compliancemanager/manager_impl.go (hunk 1 above)
Comment (dashrews78) [reply central/complianceoperator/v2/compliancemanager/manager_impl.go:241]: Yeah that is a potential race condition that would be extremely unlikely.  But if it occurred we would then track the cluster with the scan configuration again even though the objects would have been removed from sensor.  So we would be in an inconsistent state.  Thus we should probably remove any such `runningRequests` when we `processClusterDelete`

File: central/complianceoperator/v2/compliancemanager/manager_impl.go [hunk 2]
```diff
@@ -198,6 +232,38 @@ func (m *managerImpl) ProcessScanRequest(ctx context.Context, scanRequest *stora
...
+							Name: scanRequest.GetScanConfigName(),
+						},
+					},
+				},
+			},
+		}
+
+		err := m.sensorConnMgr.SendMessage(clusterID, sensorMessage)
+		if err != nil {
+			log.Errorf("error sending deletion of compliance scan config to cluster %q: %v", clusterID, err)
```
[the remark this answers is not in the document]
Comment (SimonBaeumer) [reply central/complianceoperator/v2/compliancemanager/manager_impl.go:256]: No, I don't have any. I think we are eventual consistent, so when it fails we just try again to send the same message. I think we could do it while processing Responses but doing it in-memory as of today would not make the situation  a lot better.
imho we can wait for the sync to be implemented.

File: central/complianceoperator/v2/compliancemanager/manager_impl.go (hunk 1 above)
Comment (SimonBaeumer) [reply central/complianceoperator/v2/compliancemanager/manager_impl.go:241]: In production it is unlikely. In a dev system I am seeing me running into this quite often.

Comment (SimonBaeumer) [reply central/complianceoperator/v2/compliancemanager/manager_impl.go:241]: > Thus we should probably remove any such runningRequests when we processClusterDelete

This sounds like the solution. Imho can be done in a second PR if it is faster.

Comment (dashrews78) [reply central/complianceoperator/v2/compliancemanager/manager_impl.go:241]: Will do separately.  For tracking
https://issues.redhat.com/browse/ROX-22509

**Pull request merged**